pkg/progress/progress.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package progress
+
+import (
+	"math"
+	"sync"
+	"time"
+)
+
+// Manager is used to maintain the progresses we are care about.
+type Manager struct {
+	sync.RWMutex
+	progesses map[string]*progressIndicator
+}
+
+// NewManager creates a new Manager.
+func NewManager() *Manager {
+	return &Manager{
+		progesses: make(map[string]*progressIndicator),
+	}
+}
+
+// progressIndicator reflects a specified progress.
+type progressIndicator struct {
+	total       float64
+	left        float64
+	startTime   time.Time
+	speedPerSec float64
+}
+
+// Reset resets the progress manager.
+func (m *Manager) Reset() {
+	m.Lock()
+	defer m.Unlock()
+
+	m.progesses = make(map[string]*progressIndicator)
+}
+
+// AddProgress adds a progress into manager if it doesn't exist.
+func (m *Manager) AddProgress(progress string, total float64) (exist bool) {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, exist = m.progesses[progress]; !exist {
+		m.progesses[progress] = &progressIndicator{
+			total:     total,
+			left:      total,
+			startTime: time.Now(),
+		}
+	}
+	return
+}
+
+// UpdateProgress updates a progress into manager if it doesn't exist.
+func (m *Manager) UpdateProgress(progress string, left float64) {
+	m.Lock()
+	defer m.Unlock()
+
+	if p, exist := m.progesses[progress]; exist {
+		p.left = left
+		if p.total < left {
+			p.total = left
+		}
+		p.speedPerSec = (p.total - p.left) / time.Since(p.startTime).Seconds()
+	}
+}
+
+// RemoveProgress removes a progress from manager.
+func (m *Manager) RemoveProgress(progress string) (exist bool) {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, exist = m.progesses[progress]; exist {
+		delete(m.progesses, progress)
```

Comment:
If `m.progesses` is nil or there is no such element, `delete` is a no-op. So I think there is no need to check the existence first.

Reply:
We need to return if the progress has existed to the caller.